## 분산분석

#### 책 1. 비누로 얼굴 씻기 실험
- (1) 네 가지 방법 간에 차이가 있는지 분석하시오.

In [2]:
! pip install researchpy


[notice] A new release of pip available: 22.2.2 -> 22.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
# 필요한 모듈 불러오기
import pandas as pd
import researchpy as rp
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [6]:
# 데이터 불러오기
wash = pd.read_csv("C:/Users/bluei/OneDrive - 한국방송통신대학교/문서/KNOU/2022-2/Python_R/rpython/bacteriasoap.csv")
wash.head(6)

BacterialCounts              Method
0               74               Water
1               84                Soap
2               70  Antibacterial Soap
3               51       Alcohol Spray
4              135               Water
5               51                Soap

In [25]:
# 분산분석 선형 모형 적합
fit1 = ols('wash.BacterialCounts ~ C(wash.Method)', data = wash).fit()
fit1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                             OLS Regression Results                             
================================================================================
Dep. Variable:     wash.BacterialCounts   R-squared:                       0.431
Model:                              OLS   Adj. R-squared:                  0.370
Method:                   Least Squares   F-statistic:                     7.064
Date:                  Wed, 30 Nov 2022   Prob (F-statistic):            0.00111
Time:                          21:13:41   Log-Likelihood:                -159.29
No. Observations:                    32   AIC:                             326.6
Df Residuals:                        28   BIC:                             332.4
Df Model:                             3                                         
Covariance Type:              nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                               37.5000     13.277      2.825      0.009      10.304      64.696
C(wash.Method)[T.Antibacterial Soap]    55.0000     18.776      2.929      0.007      16.539      93.461
C(wash.Method)[T.Soap]                  68.5000     18.776      3.648      0.001      30.039     106.961
C(wash.Method)[T.Water]                 79.5000     18.776      4.234      0.000      41.039     117.961
==============================================================================
Omnibus:                        4.784   Durbin-Watson:                   2.569
Prob(Omnibus):                  0.091   Jarque-Bera (JB):                3.194
Skew:                           0.638   Prob(JB):                        0.202
Kurtosis:                       3.875   Cond. No.                         4.79
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [8]:
# 일원배치 분산분석표
aov_fit1 = sm.stats.anova_lm(fit1, typ=2)
aov_fit1

sum_sq    df         F    PR(>F)
C(wash.Method)  29882.0   3.0  7.063587  0.001111
Residual        39484.0  28.0       NaN       NaN

- 평균 차이에 대한 유의확률이 0.001 로 0.05보다 작으므로, 각 수준(세정 방법) 별 평균에 차이가 없다는 귀무가설을 기각한다. 즉, 세정 방법 간에 차이가 있다고 할 수 있다.

- (2) 각 그룹 간에 차이가 있는지 다중비교를 통해 분석하시오.

In [9]:
# 필요한 모듈 불러오기
from statsmodels.stats.multicomp import pairwise_tukeyhsd
from statsmodels.stats.multicomp import MultiComparison

In [12]:
# 다중비교, Tukey 방식
mc1 = MultiComparison(wash['BacterialCounts'], wash['Method'])
mc1_tukey = mc1.tukeyhsd()
print(mc1_tukey)

             Multiple Comparison of Means - Tukey HSD, FWER=0.05              
      group1             group2       meandiff p-adj   lower    upper   reject
------------------------------------------------------------------------------
     Alcohol Spray Antibacterial Soap     55.0  0.032   3.7358 106.2642   True
     Alcohol Spray               Soap     68.5 0.0056  17.2358 119.7642   True
     Alcohol Spray              Water     79.5 0.0012  28.2358 130.7642   True
Antibacterial Soap               Soap     13.5 0.8887 -37.7642  64.7642  False
Antibacterial Soap              Water     24.5 0.5676 -26.7642  75.7642  False
              Soap              Water     11.0 0.9355 -40.2642  62.2642  False
------------------------------------------------------------------------------


- '스프레이 손에 뿌리고 씻기' 방법은 '안티박테리아 비누로 얼굴 씻기', '일반 비누로 얼굴 씻기', '물로만 얼굴 씻기' 세 가지 방법 모두와 통계적으로 유의한 차이가 있다.
- '안티박테리아 비누로 얼굴 씻기', '일반 비누로 얼굴 씻기',  '물로만 얼굴 씻기', 세 그룹은 서로 유의한 차이가 없다.

#### 책 2. 시간대별로 다운로드 하는데 걸리는 시간 측정
- (1) 세 가지 시간대별로 다운로드에 걸리는 시간이 다른지 분석하시오.

In [30]:
# 데이터 불러오기
download = pd.read_csv("C:/Users/bluei/OneDrive - 한국방송통신대학교/문서/KNOU/2022-2/Python_R/rpython/downloading.csv")
download.head(3)

TimeofDay  Time(Sec)
0  Early (7AM)         68
1  Early (7AM)        138
2  Early (7AM)         75

In [35]:
download.columns = ['td', 'sec']
download = download[['sec', 'td']]
download.head(3)

sec           td
0   68  Early (7AM)
1  138  Early (7AM)
2   75  Early (7AM)

In [36]:
# 분산분석 선형 모형 적합
fit2 = ols('download.sec ~ C(download.td)', data = download).fit()
fit2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           download.sec   R-squared:                       0.672
Model:                            OLS   Adj. R-squared:                  0.657
Method:                 Least Squares   F-statistic:                     46.03
Date:                Wed, 30 Nov 2022   Prob (F-statistic):           1.31e-11
Time:                        21:29:34   Log-Likelihood:                -251.52
No. Observations:                  48   AIC:                             509.0
Df Residuals:                      45   BIC:                             514.6
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                              113.3750     11.786      9.619      0.000      89.636     137.114
C(download.td)[T.Evening (5 PM)]       159.9375     16.668      9.595      0.000     126.366     193.509
C(download.td)[T.Late Night (12 AM)]    79.6875     16.668      4.781      0.000      46.116     113.259
==============================================================================
Omnibus:                        1.641   Durbin-Watson:                   2.265
Prob(Omnibus):                  0.440   Jarque-Bera (JB):                1.497
Skew:                           0.309   Prob(JB):                        0.473
Kurtosis:                       2.394   Cond. No.                         3.73
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [37]:
# 일원배치 분산분석표
aov_fit2 = sm.stats.anova_lm(fit2, typ=2)
aov_fit2

sum_sq    df          F        PR(>F)
C(download.td)  204640.875   2.0  46.034932  1.306432e-11
Residual        100020.125  45.0        NaN           NaN

- 평균 차이에 대한 유의확률이 1.306432^e-11 로 0.05보다 작으므로, 각 수준(시간대) 별 평균에 차이가 없다는 귀무가설을 기각한다. 즉, 시간대별로 다운로드하는 데 걸리는 시간이 차이가 있다고 할 수 있다.

- (2) 각 시간대별로 차이가 있는지 다중비교를 통해 분석하시오.

In [38]:
# 다중비교, Tukey 방식
mc2 = MultiComparison(download['sec'], download['td'])
mc2_tukey = mc2.tukeyhsd()
print(mc2_tukey)

            Multiple Comparison of Means - Tukey HSD, FWER=0.05            
    group1           group2       meandiff p-adj    lower    upper   reject
---------------------------------------------------------------------------
   Early (7AM)     Evening (5 PM) 159.9375    0.0  119.5399 200.3351   True
   Early (7AM) Late Night (12 AM)  79.6875 0.0001   39.2899 120.0851   True
Evening (5 PM) Late Night (12 AM)   -80.25    0.0 -120.6476 -39.8524   True
---------------------------------------------------------------------------


- 세 가지 시간대별로 다운로드 시간에 모두 유의한 차이가 있다.